### PR TITLE
feat(core): machine to machine apps

### PR DESCRIPTION
**Pull request by gao-sun**
## Summary

Initial support for Machine to Machine apps.

- Add a new app type “Machine to Machine”
- Add `role_names` column in applications table and application access token claims
- Only `client_credentials` grant type and `client_secret_basic` auth method are allowed for M2M apps.
- Re-group application details tabs. Move all endpoints to "Advanced settings" and add "App ID" in "Settings".

Some to-dos are pending by the sync up meeting result.

## Testing

- [x] Create a M2M app in AC, it creates the expected app and skips the guide modal.
- [x] Use Basic auth method to fetch an Access Token, decode the result and it shows correct claims

image

image

- [x] Cannot use `refresh_token` grant type for token endpoint

image

### CODE REVIEW

File: packages/schemas/tables/applications.sql [hunk 1]
```diff
@@ -8,6 +8,7 @@ create table applications (
   type application_type not null,
   oidc_client_metadata jsonb /* @use OidcClientMetadata */ not null,
   custom_client_metadata jsonb /* @use CustomClientMetadata */ not null default '{}'::jsonb,
+  role_names jsonb /* @use RoleNames */ not null default '[]'::jsonb,
```
Comment (wangsijie) [inline packages/schemas/tables/applications.sql:11]: I created LOG-4220 and will add a migration file for this.

**Review by wangsijie — COMMENTED** [review]
Then the "sub" is not a userId, do we need to check on this, like: `findUserById(ctx.req.auth)`

**Comment by gao-sun** [comment]
> Then the "sub" is not a userId, do we need to check on this, like: `findUserById(ctx.req.auth)`

Good call. We do need to update this behavior.

File: packages/schemas/tables/applications.sql (hunk 1 above)
Comment (gao-sun) [reply packages/schemas/tables/applications.sql:11]: Cool, I can do it

**Review by wangsijie — APPROVED** [review]

File: packages/core/src/oidc/adapter.ts [hunk 2]
```diff
@@ -68,10 +66,7 @@ export default function postgresAdapter(modelName: string): ReturnType<AdapterFa
       client_id,
       client_secret,
       client_name,
-      application_type: getApplicationTypeString(type),
-      grant_types: Object.values(GrantType),
-      token_endpoint_auth_method:
-        type === ApplicationType.Traditional ? 'client_secret_basic' : 'none',
+      ...getConstantClientMetadata(type),
```
Comment (simeng-li) [inline packages/core/src/oidc/adapter.ts:69]: https://github.com/panva/node-oidc-provider/blob/main/docs/README.md#clients

Can't recall. Is there any specific response we did not put `application_type`, `grant_types` and `token_endpoint_auth_method` in DB as part of the OIDC client metadata?

**Review by simeng-li — APPROVED** [review]
overall LGTM

File: packages/console/src/pages/ApplicationDetails/components/Settings.tsx [hunk 3]
```diff
@@ -74,99 +79,94 @@ const Settings = ({ applicationType, oidcConfig, defaultData, isDeleted }: Props
...
+                    ? t('application_details.redirect_uri_placeholder_native')
+                    : t('application_details.redirect_uri_placeholder')
+                }
+                onChange={onChange}
+              />
+            )}
+          />
+        </FormField>
+      )}
+      {applicationType !== ApplicationType.MachineToMachine && (
```
Comment (charIeszhao) [inline packages/console/src/pages/ApplicationDetails/components/Settings.tsx:115]: The block above is also for application type not equals to M2M, can combine these condition blocks together

File: packages/console/src/pages/ApplicationDetails/index.tsx [hunk 4]
```diff
@@ -144,13 +144,16 @@ const ApplicationDetails = () => {
...
             </div>
             <div className={styles.operations}>
-              <Button
-                title="application_details.check_guide"
-                size="large"
-                onClick={() => {
-                  setIsReadmeOpen(true);
-                }}
-              />
+              {/* TODO: @Charles figure out a better way to check guide availability */}
```
Comment (charIeszhao) [inline packages/console/src/pages/ApplicationDetails/index.tsx:147]: nit: I think we should have had guide for all application types. So IMO providing a guide for M2M should be good enough?

Linear task added to create M2M guide
https://linear.app/silverhand/issue/LOG-4228/providing-integration-guide-for-m2m-application

File: packages/phrases/src/locales/en/translation/admin-console/application-details.ts [hunk 5]
```diff
@@ -28,7 +29,7 @@ const application_details = {
   add_another: 'Add Another',
   id_token_expiration: 'ID Token expiration',
   refresh_token_expiration: 'Refresh Token expiration',
-  token_endpoint: 'Token endpoint',
+  token_endpoint: 'Token Endpoint',
```
Comment (charIeszhao) [inline packages/phrases/src/locales/en/translation/admin-console/application-details.ts:32]: Why do we need to change the letter case here? If we do, we should also change it for others, like the one below "Userinfo endpoint", right?

**Review by charIeszhao — APPROVED** [review]
Tested in GitPod. Looks good!

File: packages/core/src/oidc/adapter.ts (hunk 2 above)
Comment (gao-sun) [reply packages/core/src/oidc/adapter.ts:69]: We want to hard-code them to fit the best practice. So I think no need to store in DB

File: packages/console/src/pages/ApplicationDetails/components/Settings.tsx (hunk 3 above)
Comment (gao-sun) [reply packages/console/src/pages/ApplicationDetails/components/Settings.tsx:115]: it's by design - I tried fragment but the code looks ugly

File: packages/console/src/pages/ApplicationDetails/index.tsx (hunk 4 above)
Comment (gao-sun) [reply packages/console/src/pages/ApplicationDetails/index.tsx:147]: yes, that's in the to-do list. thanks

File: packages/phrases/src/locales/en/translation/admin-console/application-details.ts (hunk 5 above)
Comment (gao-sun) [reply packages/phrases/src/locales/en/translation/admin-console/application-details.ts:32]: because it's a term - and yes we should change all of them

Comment (gao-sun) [reply packages/phrases/src/locales/en/translation/admin-console/application-details.ts:32]: created LOG-4231

**Pull request merged**